plugins/inputs/kafka_consumer/kafka_consumer.go

```diff
@@ -143,9 +151,138 @@ func (k *KafkaConsumer) Init() error {
+	if k.topicClient == nil {
+		client, err := sarama.NewClient(k.Brokers, k.config)
+		if err != nil {
+			return err
+		}
+		k.topicClient = client
+	}
+	return nil
+}
+
+func (k *KafkaConsumer) compileTopicRegexps() {
+	// While we can add new topics matching extant regexps, we can't
+	// update that list on the fly.  We compile them once at startup.
+	// Changing them is a configuration change and requires a restart.
+
+	if len(k.TopicRegexps) == 0 {
+		// Nothing to do.
+		return
+	}
+	k.regexps = make([]regexp.Regexp, 0, len(k.TopicRegexps))
+	// Let's optimistically assume that all of them compile
+	for _, r := range k.TopicRegexps {
+		re, err := regexp.Compile(r)
+		if err == nil {
+			k.regexps = append(k.regexps, *re)
+			k.Log.Infof("Added regular expression '%s' to topics", r)
+		} else {
+			k.Log.Errorf("Regular expression '%s' did not compile: '%w'; not adding to topics", r, err)
+		}
+	}
+}
+
+func (k *KafkaConsumer) refreshTopics() error {
+	// We have instantiated a new generic Kafka client, so we can ask
+	// it for all the topics it knows about.  Then we build
+	// regexps from our strings, loop over those, loop over the
+	// topics, and if we find a match, add that topic to
+	// out topic set, which then we turn back into a list at the end.
+
+	if len(k.regexps) == 0 {
+		return nil
+	}
+	if k.topicClient == nil {
+		err := k.createTopicScanner()
+		if err != nil {
+			return fmt.Errorf("could not create client to match topics to regexps: %w", err)
+		}
+	}
+
+	allDiscoveredTopics, err := k.topicClient.Topics()
+	if err != nil {
+		return err
+	}
+	k.Log.Infof("discovered topics: %v", allDiscoveredTopics)
+
+	extantTopicSet := make(map[string]bool, len(allDiscoveredTopics))
+	for _, t := range allDiscoveredTopics {
+		extantTopicSet[t] = true
+	}
+	// Even if a topic specified by a literal string (that is, k.Topics)
+	// does not appear in the topic list, we want to keep it around, in
+	// case it pops back up--it is not guaranteed to be matched by any
+	// of our regular expressions.  Therefore, we pretend that it's in
+	// extantTopicSet, even if it isn't.
+	//
+	// Assuming that literally-specified topics are usually in the topics
+	// present on the broker, this should not need a resizing (although if
+	// you have many topics that you don't care about, it will be too big)
+	wantedTopicSet := make(map[string]bool, len(allDiscoveredTopics))
+	for _, t := range k.Topics {
+		// Get our pre-specified topics
+		k.Log.Infof("adding literally-specified topic %s", t)
+		wantedTopicSet[t] = true
+	}
+	for t := range extantTopicSet {
+		// Add topics that match regexps
+		for _, r := range k.regexps {
+			if r.MatchString(t) {
+				wantedTopicSet[t] = true
+				k.Log.Infof("adding regexp-matched topic '%s'", t)
+				break
+			}
+		}
+	}
+	topicList := make([]string, 0, len(wantedTopicSet))
+	for t := range wantedTopicSet {
+		topicList = append(topicList, t)
+	}
+	sort.Strings(topicList)
+	k.replaceTopics(topicList)
```

Comment:
Can you please elaborate why you not just overwrite the list?

Reply:
Sure: that's the comment at the top of replaceTopics():

```
        // We're not just using reflect.DeepEqual because it's slow.
        // This is pretty straightforward: we replace unless the old list
        // and the new list have all the same members in the same order.  We
        // keep them sorted internally for display purposes anyway.
        // This is slower than just replacing the topic list every time,
        // but it's nice to have a log of times when the topic list changes,
        // so we're going to want to do that comparison anyway.
```

Reply:
That could use a little rewording.

Reply:
```
        // From a functional standpoint, we could just replace the list
        // each time.
        //
        // This is a little slower, but the motivation is that we'd like
        // to log the event when the topic list changes, because it means
        // we detected a change in our matched topics (either new topics
        // appeared, or old ones went away).
        //
        // reflect.DeepEqual is said to be slow, so we implement our own
        // list-detection change.
        //
        // This is pretty straightforward: we replace unless the old list
        // and the new list have all the same members in the same order.  We
        // keep them sorted internally for display purposes anyway, so we
        // can make the decision as soon as we know the lists' lengths differ
        // or we find a different topic than before at a given index.
```

Reply:
I'm still not convinced this is necessary. Why not
```suggestion
	fingerprint := strings.Join(topicList, ";")
	if fingerprint != k.fingerprint {
		k.Log.Infof("updating topics: replacing '%v' with '%v'", k.allWantedTopics, topicList)
	}
	k.fingerprint = fingerprint
	k.allWantedTopics = topicList
```
You might alternatively use a hash-function to generate the fingerprint... This is _much_ less code and easier to understand IMO.

Reply:
I don't think there's much point in a hash function, because this is going to be...maybe hundreds of topics, rather than millions (well, I think, anyway).  The naive fingerprint is fine.